packages/zql/src/ivm/view-apply-change.ts

```diff
@@ -233,55 +239,59 @@ export function applyChange(
       if (singular) {
         const existing = parentEntry[relationship];
         assertRCEntry(existing);
-        const rc = existing[refCountSymbol];
-        const newEntry = {
-          ...existing,
-          ...change.node.row,
-          [refCountSymbol]: rc,
-        };
-        existing[refCountSymbol] = 0;
-        (parentEntry as Writable<Entry>)[relationship] = newEntry;
+        applyEdit(existing, change, schema, withIDs);
       } else {
         const view = getChildEntryList(parentEntry, relationship);
         // If the order changed due to the edit, we need to remove and reinsert.
-        if (schema.compareRows(change.oldNode.row, change.node.row) === 0) {
-          const {pos, found} = binarySearch(
+        if (schema.compareRows(change.oldNode.row, change.node.row) !== 0) {
+          // Remove
+          const {pos: oldPos, found: oldFound} = binarySearch(
             view,
             change.oldNode.row,
             schema.compareRows,
           );
-          assert(found, 'node does not exist');
-          const oldEntry = view[pos];
-          const rc = oldEntry[refCountSymbol];
-          oldEntry[refCountSymbol] = 0;
-
-          const newEntry = makeEntryPreserveRelationships(
+          assert(oldFound, 'old node does not exist');
+          const oldEntry = view[oldPos];
+          const {pos, found} = binarySearch(
+            view,
             change.node.row,
-            oldEntry,
-            format.relationships,
-            schema,
-            withIDs,
-            rc,
+            schema.compareRows,
           );
-
-          view[pos] = newEntry;
+          if (
```

Comment:
I think this needs comments explaining what is going on here. I have this fresh in my head from last night but I do not think I will be able to understand this in 6 months.